substrate/runtime/session/src/lib.rs

```diff
@@ -156,17 +178,15 @@ impl<T: Trait> Module<T> {
 		// check block number and call next_session if necessary.
 		let block_number = <system::Module<T>>::block_number();
 		let is_final_block = ((block_number - Self::last_length_change()) % Self::length()).is_zero();
-		let broken_validation = Self::broken_validation();
-		if let Some(normal_rotation) = Self::forcing_new_session() {
-			Self::rotate_session(normal_rotation, is_final_block);
-			<ForcingNewSession<T>>::kill();
-		} else if is_final_block || broken_validation {
-			Self::rotate_session(!broken_validation, is_final_block);
+		let bad_validators = <BadValidators<T>>::take();
+		let should_end_session = <ForcingNewSession<T>>::take().is_some() || bad_validators.is_some() || is_final_block;
```

Comment:
to avoid early exit on empty `bad_validators` should be `bad_validators.map_or(false, |v| !is_empty())`